webapp/channels/src/packages/mattermost-redux/src/reducers/entities/channel_bookmarks.ts

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {combineReducers} from 'redux';
+
+import type {ChannelBookmark, ChannelBookmarksState} from '@mattermost/types/channel_bookmarks';
+import type {Channel, ServerChannelWithBookmarks} from '@mattermost/types/channels';
+import type {IDMappedObjects} from '@mattermost/types/utilities';
+
+import {ChannelBookmarkTypes, UserTypes, ChannelTypes} from 'mattermost-redux/action_types';
+import type {GenericAction} from 'mattermost-redux/types/actions';
+
+const toNewObj = <T extends {id: string}>(current: IDMappedObjects<T>, arr: T[]) => {
+    return arr.reduce((acc, x) => {
+        return {...acc, [x.id]: x};
+    }, {...current});
+};
+
+export function byChannelId(state: ChannelBookmarksState['byChannelId'] = {}, action: GenericAction) {
+    switch (action.type) {
+    case ChannelBookmarkTypes.RECEIVED_BOOKMARKS: {
+        const channelId: Channel['id'] = action.data.channelId;
+        const bookmarks: ChannelBookmark[] = action.data.bookmarks;
+
+        return {
+            ...state,
+            [channelId]: toNewObj(state[channelId], bookmarks),
+        };
+    }
+
+    case ChannelBookmarkTypes.RECEIVED_BOOKMARK: {
+        const bookmark: ChannelBookmark = action.data;
+        const {id, channel_id: channelId} = bookmark;
+
+        return {
+            ...state,
+            [channelId]: {
+                ...state[channelId],
+                [id]: {
+                    ...state[channelId]?.[id],
+                    ...bookmark,
+                },
+            },
+        };
+    }
+
+    case ChannelTypes.RECEIVED_CHANNELS: {
+        const channels: ServerChannelWithBookmarks[] = action.data;
+
+        if (!channels) {
+            return state;
+        }
+
+        return channels.reduce((nextState, channel) => {
+            if (!channel.bookmarks) {
+                return nextState;
+            }
+
+            return {...nextState, [channel.id]: toNewObj(nextState[channel.id], channel.bookmarks)};
```

Comment:
at some point, when using the `bookmarks_since`, you should check if some bookmarks should be removed from state.